{"cells":[{"cell_type":"markdown","metadata":{},"source":["# Assignment-1: Computational Genomics\n","## Task 1.2"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2024-09-01T16:41:37.787359Z","iopub.status.busy":"2024-09-01T16:41:37.786759Z","iopub.status.idle":"2024-09-01T16:41:56.709409Z","shell.execute_reply":"2024-09-01T16:41:56.707931Z","shell.execute_reply.started":"2024-09-01T16:41:37.787311Z"},"trusted":true},"outputs":[],"source":["pip install scanpy"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2024-09-01T16:41:56.711521Z","iopub.status.busy":"2024-09-01T16:41:56.711192Z","iopub.status.idle":"2024-09-01T16:43:52.519830Z","shell.execute_reply":"2024-09-01T16:43:52.518699Z","shell.execute_reply.started":"2024-09-01T16:41:56.711486Z"},"trusted":true},"outputs":[],"source":["import subprocess\n","subprocess.run('conda install -c conda-forge r-base', shell=True)"]},{"cell_type":"code","execution_count":null,"metadata":{"execution":{"iopub.execute_input":"2024-09-01T16:43:52.521351Z","iopub.status.busy":"2024-09-01T16:43:52.520996Z"},"trusted":true},"outputs":[],"source":["!pip install rpy2"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["!pip install anndata2ri"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["import scanpy as sc \n","import pandas as pd\n","import numpy as np\n","import matplotlib.pyplot as plt\n","import matplotlib.ticker as ticker\n","import logging\n","import seaborn as sns\n","import anndata2ri\n","import rpy2.rinterface_lib.callbacks as rcb\n","import rpy2.robjects as ro\n","\n","rcb.logger.setLevel(logging.ERROR)\n","ro.pandas2ri.activate()\n","anndata2ri.activate()\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix = sc.read('/kaggle/input/dataset.h5ad')\n","matrix"]},{"cell_type":"markdown","metadata":{},"source":["What does this output mean? \n","- n_obs refers to the number of cells. Hence, here we have 10,727 cells. \n","- n_vars refers to the number of genes. This matrix contains details about 12,303 genes. \n","\n","Inspecting the AnnData object. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix.var_names"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix.obs_names"]},{"cell_type":"markdown","metadata":{},"source":["Adding the gene IDs provided."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["gene_IDs = pd.read_csv('/kaggle/input/gene_names.csv')\n","gene_IDs"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix.var_names = gene_IDs['0'].values\n","matrix.var_names"]},{"cell_type":"markdown","metadata":{},"source":["### Quality Control\n","Finding the mitochondrial, ribosomal and hemoglobin genes. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# mitochondrial genes\n","matrix.var[\"mt\"] = matrix.var_names.str.startswith(\"MT-\")\n","# ribosomal genes\n","matrix.var[\"ribo\"] = matrix.var_names.str.startswith((\"RPS\", \"RPL\"))\n","# hemoglobin genes\n","matrix.var[\"hb\"] = matrix.var_names.str.contains(\"^HB[^(P)]\")\n","sc.pp.calculate_qc_metrics(\n","    matrix, qc_vars=[\"mt\", \"ribo\", \"hb\"], inplace=True, log1p=True\n",")"]},{"cell_type":"markdown","metadata":{},"source":["The function above, adds the qc metrics to our AnnData object. The log1p=True returns the logarithm of the results (log(1+x)), which accounts for skewed distrubutions. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pl.violin(\n","    matrix,\n","    [\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n","    jitter=0.4,\n","    multi_panel=True,\n",")"]},{"cell_type":"markdown","metadata":{},"source":["It is clear from the above violin plots that there are no mitochondrial genes expressed. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pl.scatter(matrix, \"total_counts\", \"n_genes_by_counts\")"]},{"cell_type":"markdown","metadata":{},"source":["To choose apt quality control thresholds, it may be helpful to draw histograms. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# Plot a histogram of total counts\n","plt.hist(matrix.obs['total_counts'], bins=100, color='skyblue', edgecolor='black')\n","plt.title('Total Counts per Cell')\n","plt.xlabel('Total Counts')\n","plt.ylabel('Number of Cells')\n","plt.show()"]},{"cell_type":"markdown","metadata":{},"source":["There seems to be a noise peak in the above plot below the value of 1400 total counts. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# Plot a histogram of number of genes per barcode\n","plt.hist(matrix.obs['n_genes_by_counts'], bins=100, color='skyblue', edgecolor='black')\n","plt.title('Number of Unique Genes Expressed Per Cell')\n","plt.xlabel('Number of Unique Genes')\n","plt.ylabel('Number of Cells')\n","plt.show()"]},{"cell_type":"markdown","metadata":{},"source":["In this graph there is no apparent noise peak on the lower end. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["TotalCounts = matrix.obs['total_counts'].values\n","sorted_counts = np.sort(TotalCounts)[::-1]\n","ranks = np.arange(1, len(sorted_counts) + 1)\n","plt.figure(figsize=(8, 6))\n","plt.plot(ranks, sorted_counts, marker='o', linestyle='none', color='blue', markersize=1)\n","plt.yscale('log')\n","plt.xlabel('Rank of Cells (High to Low)')\n","ax = plt.gca()\n","ax.yaxis.set_major_locator(ticker.LogLocator(base=10.0, subs='auto'))\n","ax.yaxis.set_minor_locator(ticker.LogLocator(base=10.0, subs='auto', numticks=10))\n","ax.yaxis.set_major_formatter(ticker.LogFormatter(base=10.0))\n","plt.ylabel('Total Counts (log scale)')\n","plt.title('Count Depth Distribution (Log-Log Plot)')\n","plt.grid(True, which='both', linestyle='--', linewidth=0.7)\n","plt.show()"]},{"cell_type":"markdown","metadata":{},"source":["In the above graph, the \"Elbow\" is around 1600 total counts."]},{"cell_type":"markdown","metadata":{},"source":["### Doublet Detection\n","Making use of Scrublet - the nearest neighbour based doublet detection algorithm."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pp.scrublet(matrix)\n","np.sum(matrix.obs['predicted_doublet'])"]},{"cell_type":"markdown","metadata":{},"source":["Hence, 161 doublets have been detected."]},{"cell_type":"markdown","metadata":{},"source":["### Normalization\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["#Saving the raw counts\n","matrix.layers[\"counts\"] = matrix.X.copy()\n","# Normalizing to median total counts and transforming it to log(1+x)\n","sc.pp.normalize_total(matrix)\n","sc.pp.log1p(matrix)"]},{"cell_type":"markdown","metadata":{},"source":["### Feature Selection\n","This step is to reduce the dimensions of the matrix such that only the most informative genes are retained. Scanpy automatically annotates the highly variable genes using variance to mean ratios. Usually, erring towards the removal of a larger number of highly variable genes often produces the best results. Hence, here we choose the top 2000 most variable genes.   "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["\n","%load_ext rpy2.ipython"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%R\n","if (!requireNamespace(\"BiocManager\", quietly = TRUE)) {\n","    install.packages(\"BiocManager\", quiet = TRUE, repos = \"http://cran.us.r-project.org\")\n","}\n","\n","#installing the required packages\n","BiocManager::install(\"S4Vectors\", ask = FALSE, update = FALSE, quiet = TRUE)\n","BiocManager::install(\"scry\", ask = FALSE, update = FALSE, quiet = TRUE)\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["%%R\n","library(scry)\n","\n","sce = devianceFeatureSelection(matrix, assay=\"X\")\n","\n","binomial_deviance = rowData(sce)$binomial_deviance\n","\n","idx = order(binomial_deviance, decreasing = TRUE)[1:2000]\n","mask = rep(FALSE, length(binomial_deviance))\n","mask[idx] = TRUE\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["mask = ro.globalenv['mask']\n","binomial_deviance = ro.globalenv['binomial_deviance']"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# Back to Python\n","# Update the AnnData object with highly deviant genes\n","matrix.var[\"highly_deviant\"] = mask\n","matrix.var[\"binomial_deviance\"] = binomial_deviance"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix"]},{"cell_type":"markdown","metadata":{},"source":["As visible, the 'highly_variable' and 'binomial_deviance' key has now been added to var. "]},{"cell_type":"markdown","metadata":{},"source":["### Dimensionality Reduction\n","Performing Principal Component Analysis (PCA)."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix = matrix.copy()\n","sc.tl.pca(matrix)\n","sc.pl.pca_variance_ratio(matrix, n_pcs=50, log=True)\n"]},{"cell_type":"markdown","metadata":{},"source":["The \"elbow\" of the above curve is around the region of 10-11 principal components. Since there are no pronounced disadvantages of overestimating the principal components, we can fix the value at 50 principal components."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pp.pca(matrix, n_comps = 50)"]},{"cell_type":"markdown","metadata":{},"source":["### Constructing the Nearest Neighbours Graph\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pp.neighbors(matrix)\n","sc.tl.umap(matrix)\n","sc.pl.umap(matrix, size = 2,)"]},{"cell_type":"markdown","metadata":{},"source":["### Visualizing the QC Metrics"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pl.umap(\n","    matrix,\n","    color=[\"predicted_doublet\", \"doublet_score\"],\n","    # increase horizontal space between panels\n","    wspace=0.5,\n","    size=3,\n",")"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pl.umap(\n","    matrix,\n","    color=[\"log1p_total_counts\", \"log1p_n_genes_by_counts\"],\n","    wspace=0.5,\n","    ncols=2,\n",")"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pl.umap(\n","    matrix,\n","    color=[\"log1p_total_counts\", \"log1p_n_genes_by_counts\"],\n","    wspace=0.5,\n","    ncols=2,\n",")"]},{"cell_type":"markdown","metadata":{},"source":["Let us now find the highly variable genes for clustering."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["#5000\n","sc.pp.highly_variable_genes(matrix, n_top_genes=5000)\n","matrix_5000 = matrix[:, matrix.var[\"highly_variable\"]]\n","sc.pp.neighbors(matrix_5000)\n","sc.tl.umap(matrix_5000)\n","sc.pl.umap(matrix_5000, color = ['total_counts'], size = 2,)\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix_5000"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["matrix = matrix_5000"]},{"cell_type":"markdown","metadata":{},"source":["# Clustering\n","Making use of leiden clustering. Let us observe how the clusters form at multiple different resolutions. "]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["for res in [0.05, 0.1, 0.20, 0.25, 1.0, 2.0]:\n","    sc.tl.leiden(\n","        matrix, key_added=f\"leiden_res_{res:4.2f}\", resolution=res, flavor=\"igraph\"\n","    )"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["sc.pl.umap(\n","    matrix,\n","    color=[\"leiden_res_0.05\", \"leiden_res_0.10\", \"leiden_res_0.20\", \"leiden_res_0.25\", \"leiden_res_1.00\", \"leiden_res_2.00\" ],\n","    legend_loc=\"on data\",\n",")"]},{"cell_type":"markdown","metadata":{},"source":["All clusters above 0.25 seem somewhat overclustered. Let us work further with the 0.25 resolution Leiden clusters."]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["#Obtaining Output Data\n","leiden_out = matrix.obs['leiden_res_0.20']\n","leiden_df = pd.DataFrame(leiden_out)\n","leiden_df.to_csv('leiden-0.20-dev.csv', index=True)\n"]},{"cell_type":"code","execution_count":null,"metadata":{"trusted":true},"outputs":[],"source":["# Obtain cluster-specific differentially expressed genes\n","sc.tl.rank_genes_groups(matrix, groupby=\"leiden_res_0.20\", method=\"wilcoxon\")\n","sc.pl.rank_genes_groups_dotplot(\n","    matrix, groupby=\"leiden_res_0.20\",standard_scale=\"var\", n_genes=5\n",")"]}],"metadata":{"kaggle":{"accelerator":"nvidiaTeslaT4","dataSources":[{"datasetId":5602417,"sourceId":9259254,"sourceType":"datasetVersion"}],"dockerImageVersionId":30762,"isGpuEnabled":true,"isInternetEnabled":true,"language":"python","sourceType":"notebook"},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.14"}},"nbformat":4,"nbformat_minor":5}
